api/webrtc/webrtc.go

```diff
@@ -0,0 +1,345 @@
+		if candidate == nil {
+			localDesc := c.pc.LocalDescription()
+			encodedDescr := Encode(localDesc)
+			c.offer = encodedDescr
+			c.OnSdp <- encodedDescr
+		}
+	})
+
+	dc, err := c.pc.CreateDataChannel(lable, nil)
+	if err != nil {
+		return err
+	}
+	dc.OnOpen(func() {
+		log.Debugf("data channel %v has been opened\n", dc.Label())
+		c.OnConnected <- true
+	})
+	dc.OnMessage(func(msg webrtc.DataChannelMessage) {
+		var dcmsg DataChannelMessage
+		if msg.IsString {
+			dcmsg.messageType = TextMessage
+			if string(msg.Data) == PongData {
+				if c.pongHandler != nil {
+					c.pongHandler(PongData)
+				} else {
+					log.Info("Pong handler not set")
+				}
+				return
+			} else if string(msg.Data) == PingData {
+				dc.SendText(PongData)
+				return
+			}
+		} else {
+			dcmsg.messageType = BinaryMessage
+		}
+		dcmsg.data = msg.Data
+		c.OnMessage <- dcmsg
+	})
+	dc.OnClose(func() {
+		c.OnConnected <- false
+	})
+
+	dc.OnError(func(err error) {
+		log.Errorf("Data Channel %s error: %s\n", dc.Label(), err.Error())
+		c.OnConnected <- false
+	})
+
+	offer, err := c.pc.CreateOffer(&webrtc.OfferOptions{ICERestart: false})
+	if err != nil {
+		return nil
+	}
+	if err = c.pc.SetLocalDescription(offer); err != nil {
+		return nil
+	}
+
+	c.dc = dc
+	return nil
+}
+
+func (c *Peer) Answer(offerSdp string) error {
+	offer := webrtc.SessionDescription{}
+	Decode(offerSdp, &offer)
+
+	sdp, err := offer.Unmarshal()
+	if err != nil {
+		return err
+	}
+	limiter := ratelimiter.GetLimiter("webrtc:"+sdp.Origin.UnicastAddress, config.Parameters.WsIPRateLimit, int(config.Parameters.WsIPRateBurst))
+	if !limiter.Allow() {
+		return fmt.Errorf("webrtc connection limit of %s reached", sdp.Origin.UnicastAddress)
+	}
+
+	if err := c.pc.SetRemoteDescription(offer); err != nil {
+		return err
+	}
+
+	answer, err := c.pc.CreateAnswer(nil)
+	if err != nil {
+		return err
+	}
+
+	if err := c.pc.SetLocalDescription(answer); err != nil {
+		return err
+	}
+
+	c.pc.OnICECandidate(func(candidate *webrtc.ICECandidate) {
+		if candidate == nil {
+			localDesc := c.pc.LocalDescription()
+			encodedDescr := Encode(localDesc)
+			c.answer = encodedDescr
+			c.OnSdp <- encodedDescr
+		}
+	})
+
+	c.pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		dc.OnOpen(func() {
+			c.dc = dc
+			c.OnConnected <- true
+			if NewConnection != nil {
+				go NewConnection(c, nil)
+			}
+		})
+
+		dc.OnMessage(func(msg webrtc.DataChannelMessage) {
+			var dcmsg DataChannelMessage
+			if msg.IsString {
+				dcmsg.messageType = TextMessage
+				if string(msg.Data) == PingData {
+					dc.SendText(PongData)
+					return
+				} else if string(msg.Data) == PongData {
+					if c.pongHandler != nil {
+						c.pongHandler(PongData)
+					} else {
+						log.Info("Pong handler not set")
+					}
+					return
+				}
+			} else {
+				dcmsg.messageType = BinaryMessage
+			}
+			dcmsg.data = msg.Data
+			c.OnMessage <- dcmsg
+		})
+
+		dc.OnClose(func() {
+			c.OnConnected <- false
```

Comment:
need to add default to prevent stuck since the chan is not read anywhere

Reply:
Yes, modified.